### PR TITLE
US123426: Support link attachments for feedback

### PR DESCRIPTION
**Pull request by ktonon**
Video and audio notes are going to be stored as links instead of files. Results in a better user experience (fewer clicks to get to the video). This add support to evaluation for handling link attachments

### CODE REVIEW

File: components/controllers/ConsistentEvaluationController.js [hunk 1]
```diff
@@ -148,22 +148,33 @@ export class ConsistentEvaluationController {
...
+				updatedEntity = await this._performAction(targetEntity,
+					saveFeedbackAttachmentFileActionName, saveFeedbackAttachmentAFieldName,
+					JSON.stringify({
+						FileId: attachment.fileId,
+						FileSystemType: attachment.fileSystemType,
+					}));
+			} else if (attachment.GetLocation) {
+				updatedEntity = await this._performAction(targetEntity,
+					saveFeedbackAttachmentLinkActionName, saveFeedbackAttachmentAFieldName,
+					JSON.stringify({
```
Comment (jmurray-d2l) [inline components/controllers/ConsistentEvaluationController.js:168]: Have we tested this value being sent?  I'm not entirely sure the API will be able to deserialize this JSON

Comment (ktonon) [reply components/controllers/ConsistentEvaluationController.js:168]: There is an upcoming PR to the lms which will handle this. The conditional branch `if (attachment.GetLocation)` should not be triggered until the lms PR is merged, as eval does not currently allow link attachments.

I will reference the lms PR once it is available

File: components/controllers/ConsistentEvaluationController.js [hunk 2]
```diff
@@ -148,22 +148,33 @@ export class ConsistentEvaluationController {
...
-			updatedEvaluationEntity = await this._performAction(targetEntity, saveFeedbackAttachmentActionName, saveFeedbackAttachmentAFieldName, value);
+			if (attachment.fileSystemType) {
+				updatedEntity = await this._performAction(targetEntity,
+					saveFeedbackAttachmentFileActionName, saveFeedbackAttachmentAFieldName,
+					JSON.stringify({
+						FileId: attachment.fileId,
+						FileSystemType: attachment.fileSystemType,
+					}));
+			} else if (attachment.GetLocation) {
+				updatedEntity = await this._performAction(targetEntity,
```
Comment (jmurray-d2l) [inline components/controllers/ConsistentEvaluationController.js:166]: nit: We have a mix of styles here with multiple params on some lines and singles others.  Consider moving each param to its own line

**Review by jmurray-d2l — APPROVED** [review]

File: components/controllers/ConsistentEvaluationController.js [hunk 3]
```diff
@@ -148,22 +148,37 @@ export class ConsistentEvaluationController {
...
-			const targetEntity = updatedEvaluationEntity.getSubEntityByRel('attachments');
+		let updatedEntity = evaluationEntity;
+		for (const attachment of files) {
+			const targetEntity = updatedEntity.getSubEntityByRel('attachments');
 			if (!targetEntity) {
 				throw new Error(ConsistentEvaluationControllerErrors.ERROR_FETCHING_ATTACHMENTS_ENTITY);
 			}
 
-			updatedEvaluationEntity = await this._performAction(targetEntity, saveFeedbackAttachmentActionName, saveFeedbackAttachmentAFieldName, value);
+			if (attachment.fileSystemType) {
```
Comment (ktonon) [inline components/controllers/ConsistentEvaluationController.js:158]: Caught this bug in manual testing. This should be `m_fileSystemType`. Going to fix it and add a unit test to catch it

Comment (jmurray-d2l) [reply components/controllers/ConsistentEvaluationController.js:158]: Same thing in the stringify?

Comment (ktonon) [reply components/controllers/ConsistentEvaluationController.js:158]: Yes, I'll all references

File: components/controllers/ConsistentEvaluationController.js [hunk 4]
```diff
@@ -148,22 +148,37 @@ export class ConsistentEvaluationController {
...
 
-			updatedEvaluationEntity = await this._performAction(targetEntity, saveFeedbackAttachmentActionName, saveFeedbackAttachmentAFieldName, value);
+			if (attachment.GetFileSystemType) {
+				updatedEntity = await this._performAction(
+					targetEntity,
+					saveFeedbackAttachmentFileActionName,
+					saveFeedbackAttachmentAFieldName,
+					JSON.stringify({
+						FileId: attachment.GetId(),
+						FileSystemType: attachment.GetFileSystemType(),
```
Comment (ktonon) [inline components/controllers/ConsistentEvaluationController.js:165]: Updated to use [D2L.LP.Web.UI.Files.File](https://search.d2l.dev/xref/lms/lp/framework/web/D2L.LP.Web/UI/Files/JavaScript/File.js?r=f06bb6c6) interface

File: test/ConsistentEvaluationController.test.js [hunk 5]
```diff
@@ -30,6 +31,53 @@ describe('ConsistentEvaluationController', () => {
...
+				fakeTargetEntity,
+				'add-link',
+				'value',
+				JSON.stringify({
+					LinkId: fakeId,
+					Name: 'Some link',
+					Url: 'http://foo',
+					Urn: 'd2l:brightspace:...'
+				})
+			]);
```
Comment (ktonon) [inline test/ConsistentEvaluationController.test.js:77]: I was worried this test would be flakey if `JSON.stringify` did not consistently order the keys, but it seems to pass every time

**Pull request merged**